### PR TITLE
feat(magi2): add fused BF16 routed MoE path

### PR DESCRIPTION
**Pull request by yeahdongcn**
## Summary

Add a fused BF16 routed-MoE path for MAGI-2 Preview and use it by default for
supported eager GPU inference. CUDA falls back to the native path below 4096
tokens, where launch overhead dominates; CPU, non-BF16, deterministic, and
`torch.compile` execution retain the native implementation.

The fast path keeps FP32 accumulation through SwiGLU7 and the down projection.
Gate/up weights are packed once after checkpoint loading and cached by
parameter/storage version. Route metadata uses reusable upper-bound buffers;
the live padded count stays device-side, avoiding per-forward allocations and
CPU/GPU synchronization.

No environment variable is required.

## Validation

```bash
python -m pytest --noconftest -o addopts='' -q \
  tests/diffusion/models/magi2/test_bf16_moe_wiring.py \
  tests/diffusion/models/magi2/test_moe_kernel_contract.py \
  tests/benchmarks/test_magi2_bf16_moe_benchmark.py \
  -m 'core_model and cpu'
```

Result on commit `aaeede2d`:

- CPU: 42 passed
- NVIDIA H20 / CUDA / Triton 3.7.1: 16 passed
- MTT S5000 / MUSA / Triton 3.2.0: 16 passed, 2 skipped
  (`BF16 atomic_add` is unavailable before Triton-MUSA 3.6; the fused path
  does not use it)
- targeted pre-commit: passed

Production-local shape (`3` EP4-local heads, `256` experts/head, top-k `6`,
hidden size `256`, intermediate size `1280`):

```bash
# H20 native atomic reference
python -m benchmarks.kernels.benchmark_magi2_bf16_moe \
  --tokens 4096 --heads 3 --experts 256 --top-k 6 \
  --warmup 10 --iters 30

# S5000 portable native reference
python -m benchmarks.kernels.benchmark_magi2_bf16_moe \
  --tokens 4096 --heads 3 --experts 256 --top-k 6 \
  --warmup 10 --iters 30 --native-deterministic
```

| Device | Native p50 | Default fused p50 | Speedup |
|---|---:|---:|---:|
| MTT S5000 60-SM, Triton 3.2.0 | 12.357 ms | 4.957 ms | 2.493x |
| NVIDIA H20, Triton 3.7.1 | 3.320 ms | 2.377 ms | 1.397x |

Both runs passed numerical parity. Maximum absolute difference was
`0.001953125` on S5000 and `0` on H20. The benchmark uses alternating
native/default calls with a 512 MiB cache flush outside timed regions; it is an
eager operator diagnostic rather than a full-video throughput claim.

## Scope

This PR contains the BF16 routed-MoE kernels, MAGI-2 wiring, focused tests, and
the reproducible microbenchmark. It does not include EP framework changes,
attention, mHC, SwiGLU7 activation, sampler changes, or MUSA-only APIs. #7156
remains the integration reference.

### CODE REVIEW

**Comment by hsliuustc0106** [comment]
This PR touches **tests/diffusion/, vllm_omni/diffusion/** (4 files). Based on CODEOWNERS coverage of the changed files, the most-related reviewers appear to be:

> @Bounty-hunter @fhfuih @wtomin

Could one of you take a look when you get a chance? Thanks!

File: vllm_omni/diffusion/models/magi2/mh_moe.py [hunk 1]
```diff
@@ -492,6 +301,13 @@ def _route(self, x_heads: torch.Tensor) -> tuple[torch.Tensor, torch.Tensor]:
 
     def _local_forward(self, x_heads: torch.Tensor) -> torch.Tensor:
         probabilities, indices = self._route(x_heads)
+        if (
+            os.environ.get("MAGI2_USE_BF16_MOE_KERNEL", "0") == "1"
```
Comment (hsliuustc0106) [inline vllm_omni/diffusion/models/magi2/mh_moe.py:305]: [P2] `MAGI2_USE_BF16_MOE_KERNEL` is missing from `vllm_omni/config/environment_variable_inventory.py`, so `tests/config/test_environment_variables.py::test_statically_resolvable_environment_accesses_are_classified` fails at this head (verified on a clean checkout: 1 failed, 12 passed; cpu/core_model lane). Please register it like the other `MAGI2_*` names and bump the pinned category counts in `test_inventory_matches_reviewed_snapshot_counts`.

File: vllm_omni/diffusion/models/magi2/mh_moe.py [hunk 2]
```diff
@@ -77,302 +73,115 @@ def compute_topk_probs_and_indices(
...
+        torch.arange(num_heads, device=x_heads.device, dtype=torch.int32).view(num_heads, 1, 1) * experts_per_head
+    )
+    route_ids = (indices.to(torch.int32) + head_offsets).reshape(num_heads * num_tokens, top_k)
+    route_weights = probabilities.reshape(num_heads * num_tokens, top_k).contiguous()
+    sorted_ids, expert_ids, num_padded = _align_bf16_routes(route_ids, num_experts, 128)
+
+    hidden = x_heads.permute(1, 0, 2).contiguous().reshape(num_heads * num_tokens, hidden_size)
+    intermediate_size = w_gate.shape[-1]
+    # [E, I, 2, D] -> [E, 2I, D]: adjacent rows are gate/up pairs.
+    packed_w13 = torch.stack((w_gate.transpose(1, 2), w_up.transpose(1, 2)), dim=2).reshape(
```
Comment (hsliuustc0106) [inline vllm_omni/diffusion/models/magi2/mh_moe.py:131]: [P2→P3 — updated after an e2e A/B] `packed_w13` is re-stacked from gate/up on every forward: at real dims (12 heads × 256 experts) that is a 4.03 GB copy per layer call (8.85 ms p50 standalone on L20X), and it makes the op-level dispatch benchmark a net regression at ≤10k-token shapes (512 tok 0.28×, 4096 0.49×, 10000 0.75×). A full e2e A/B (272p 448×256, 10 steps, ulysses SP4, seed 42, 7 runs/arm on L20X) then showed the opt-in path at e2e **parity**: clean-window generation 44–52 s for both arms (on ≈48.5 s vs off ≈46.9 s; both arms hit a ~2× slower co-tenant-interference class uncorrelated with the arm). Two bonuses the PR doesn't claim: the path is run-to-run **deterministic** — bit-identical mp4 md5 across reps and across two different hosts — and on-vs-off divergence (26.1–26.6 dB PSNR) sits inside the reference path's own atomic-nondeterminism envelope (25.5 dB run-to-run). So the repack is pure wasted device traffic (~4 GB × layers × steps per run), not a blocking regression: please cache the packed weights per weight version (or precompute at load when the flag is on) and add a real-dims row to the op-level table. The env-registry omission (other comment) remains the merge blocker.

File: vllm_omni/diffusion/models/magi2/fused_moe_kernels.py [hunk 3]
```diff
@@ -0,0 +1,566 @@
...
+        gate = expert_input @ w_gate[flat_expert]
+        up = expert_input @ w_up[flat_expert]
+        hidden = swiglu7_pair(gate, up)
+        expert_output = hidden @ w_down[flat_expert]
+        expert_output = expert_output * probs[begin:end, None].to(expert_output.dtype)
+        output[:, head].index_add_(0, token_ids, expert_output)
+    return output
+
+
+# Triton 3.7 keeps ``tl.constexpr`` as an annotation/type marker, not a
```
Comment (hsliuustc0106) [inline vllm_omni/diffusion/models/magi2/fused_moe_kernels.py:323]: [P3] This comment says Triton 3.7's `tl.constexpr` is an annotation marker rather than a callable constructor, but the code two lines below still calls `tl.constexpr(1.702)` — it reads stale from an earlier iteration; please reword or drop it.

**Review by hsliuustc0106 — CHANGES_REQUESTED** [review]
Two P2s block this at the current head:

1. `MAGI2_USE_BF16_MOE_KERNEL` is unregistered in `environment_variable_inventory.py` — `tests/config/test_environment_variables.py::test_statically_resolvable_environment_accesses_are_classified` fails on a clean checkout of 977f59a, so any cpu/core_model lane will go red.

2. The per-call `packed_w13` repack makes the opt-in path a net regression at production expert counts. Dispatch-mode A/B on L20X (12 heads × 256 experts, top-k 6, env toggled in one process): 512 tok 3374→11963 µs (0.28×), 4096 tok 6443→13137 µs (0.49×), 10000 tok 12832→17054 µs (0.75×). The stack alone measures 8.85 ms p50 (4.03 GB) per call at those dims. The reported 7.9–11.2× reproduce only at the benchmark's synthetic default `--experts 8`.

The rest verifies clean on my side: 44 CPU + 16 CUDA tests pass, parity is exact (rel-L2 0.0 at every shape including real dims), routing weights are applied exactly once in FP32, and the deterministic/env gating is correct. Packing once per weight version plus a real-dims row in the perf table should get this approve-able.

File: vllm_omni/diffusion/models/magi2/mh_moe.py [hunk 4]
```diff
@@ -77,302 +73,115 @@ def compute_topk_probs_and_indices(
...
+    order = torch.argsort(flat_ids)
+    sorted_experts = flat_ids[order]
+    counts = torch.bincount(flat_ids, minlength=num_experts)
+    padded_counts = ((counts + block_size - 1) // block_size) * block_size
+    starts = torch.cumsum(padded_counts, 0) - padded_counts
+    ends = torch.cumsum(counts, 0)
+    begins = ends - counts
+    positions = torch.arange(route_count, device=flat_ids.device, dtype=torch.int64)
+    destinations = starts[sorted_experts.long()] + positions - begins[sorted_experts.long()]
+    total = int(padded_counts.sum().item())
```
Comment (lishunyang12) [inline vllm_omni/diffusion/models/magi2/mh_moe.py]: Preallocate route buffers and keep padded counts on GPU; don't call `.item()` here.

File: vllm_omni/diffusion/models/magi2/mh_moe.py (hunk 1 above)
Comment (yeahdongcn) [reply vllm_omni/diffusion/models/magi2/mh_moe.py]: Addressed by removing the opt-in switch entirely. Supported eager BF16 GPU shapes now use the fused path by default, while CPU, non-BF16, deterministic, torch.compile, and small CUDA shapes retain the native path. There is no MAGI2_USE_BF16_MOE_KERNEL access left, so no inventory entry is needed.

File: vllm_omni/diffusion/models/magi2/mh_moe.py (hunk 2 above)
Comment (yeahdongcn) [reply vllm_omni/diffusion/models/magi2/mh_moe.py]: Addressed. W_gate/W_up are packed after checkpoint loading and cached by parameter object, storage pointer, and version; reloads and device moves invalidate the non-persistent cache. The production-local 3-head x 256-expert benchmark now measures the cached default path: S5000 12.357 ms -> 4.957 ms (2.493x) and H20 3.329 ms -> 2.371 ms (1.404x) at 4096 tokens.

File: vllm_omni/diffusion/models/magi2/fused_moe_kernels.py (hunk 3 above)
Comment (yeahdongcn) [reply vllm_omni/diffusion/models/magi2/fused_moe_kernels.py]: Addressed. The stale Triton 3.7 explanation was replaced with an accurate compatibility comment for Triton 3.2/3.7 and vLLM CPU placeholders; the constexpr constants remain guarded by HAS_TRITON.

File: vllm_omni/diffusion/models/magi2/mh_moe.py (hunk 4 above)
Comment (yeahdongcn) [reply vllm_omni/diffusion/models/magi2/mh_moe.py]: Addressed. Route metadata uses reusable upper-bound buffers, and the live padded count remains a device tensor consumed by the kernel. The per-forward `.item()` and exact-size allocations are removed; tests cover reuse and resize invalidation.

**Review by hsliuustc0106 — APPROVED** [review]
Both blockers are resolved at this head, and the env-var one the better way: instead of registering `MAGI2_USE_BF16_MOE_KERNEL`, the flag is gone entirely — the fused path now self-selects (MUSA always; CUDA at ≥4096 tokens; still respects the registered `MAGI2_DETERMINISTIC` and skips under torch.compile), so the static-classification test passes by construction. The per-forward `packed_w13` re-stack is replaced by `_get_bf16_packed_w13()` with an id/data_ptr/_version invalidation key that correctly handles in-place reloads and mmap swaps, and the stale Triton constexpr comment is fixed. Mergeable, GitHub checks green; the kernel itself is unchanged from the A/B-verified revision.
